### PR TITLE
fix OR handling in subquery decorrelation

### PR DESCRIPTION
**Pull request by 9DemonFox**
Fix issue 213

### CODE REVIEW

**Comment by zhouqingqing** [comment]
don't forget to fix codefactor issues

File: qpmodel/LogicNode.cs [hunk 1]
```diff
@@ -38,6 +38,8 @@
 
 using LogicSignature = System.Int64;
 using BitVector = System.Int64;
+using Microsoft.Scripting.Utils;
```
Comment (zhouqingqing) [inline qpmodel/LogicNode.cs:41]: do you need them?

File: qpmodel/subquery.cs [hunk 2]
```diff
@@ -92,7 +94,23 @@ public partial class SelectStmt : SQLStatement
...
+            {
+                return hasAnyExtreSubqueryExprInOR(xOR.rchild_(), exsitsExpr) || hasAnyExtreSubqueryExprInOR(xOR.lchild_(), exsitsExpr);
+            }
+            else
+            {
+                return x.VisitEachExists(y => (y is SubqueryExpr && (!y.Equals(exsitsExpr))));
+            }
+        }
+
+        // 如果有or， 就不能变成filter，否则无法替换
```
Comment (zhouqingqing) [inline qpmodel/subquery.cs:112]: English comments only please

File: qpmodel/LogicNode.cs [hunk 3]
```diff
@@ -686,11 +703,60 @@ public override List<int> ResolveColumnOrdinal(in List<Expr> reqOutput, bool rem
             var rtables = rchild_().InclusiveTableRefs();
             var lreq = new HashSet<Expr>();
             var rreq = new HashSet<Expr>();
-
-            Expr thisReq = null;
+            // seperate the output to related child respectively
+            Debug.Assert((this.lchild_() != null) && (this.rchild_() != null));
+            if (this is LogicMarkJoin LMJ)
```
Comment (zhouqingqing) [inline qpmodel/LogicNode.cs:708]: not sure I can follow the logic here - comments?

**Comment by 9DemonFox** [comment]
The company network has some problem to access to github. I will push the doc and code next monday.

**Comment by 9DemonFox** [comment]
required comments are added. But the complex codes issue confuse me a lot.  I try a lot to simplefy it.

**Comment by 9DemonFox** [comment]
image
When I open the file in CodeFactor，it will jump to file in **3c8f277**, but the update file is in **01ea6b6**. I'm not sure if there is any problem in CodeFactor checker.

**Comment by zhouqingqing** [comment]
> But the complex codes issue confuse me a lot.

This one is ok - we have multiple issues like this.

**Pull request merged**